File: benchmarks/cross_language/c/enum_match.c
/* Benchmark: Enum dispatch -- C equivalent.
 * Tagged union with enum discriminant + union payload, switch-based dispatch.
 * 6 variants matching Mapanare's Shape: Circle, Square, Triangle, Point, Line, Rect.
 * Expected: checksum = 52818168.
 */
#include <stdio.h>
#include <stdint.h>
#include <time.h>
#include <sys/resource.h>
#include <sys/time.h>

typedef enum {
    TAG_CIRCLE = 0,
    TAG_SQUARE = 1,
    TAG_TRIANGLE = 2,
    TAG_POINT = 3,
    TAG_LINE = 4,
    TAG_RECT = 5
} ShapeTag;

typedef struct {
    ShapeTag tag;
    union {
        int64_t radius;      /* Circle */
        int64_t side;        /* Square */
        struct { int64_t base, height; } tri;  /* Triangle */
        int64_t length;      /* Line */
        struct { int64_t w, h; } rect;  /* Rect */
    } payload;
} Shape;

static int64_t area(Shape s) {
    switch (s.tag) {
        case TAG_CIRCLE:   return s.payload.radius * s.payload.radius * 3;
        case TAG_SQUARE:   return s.payload.side * s.payload.side;
        case TAG_TRIANGLE: return s.payload.tri.base * s.payload.tri.height / 2;
        case TAG_POINT:    return 0;
        case TAG_LINE:     return s.payload.length;
        case TAG_RECT:     return s.payload.rect.w * s.payload.rect.h;
    }
    return 0; /* unreachable */
}

static Shape make_shape(int64_t i) {
    Shape s;
    int64_t tag = i % 6;
    switch (tag) {
        case 0: s.tag = TAG_CIRCLE;   s.payload.radius = i % 50 + 1; break;
        case 1: s.tag = TAG_SQUARE;   s.payload.side = i % 30 + 1; break;
        case 2: s.tag = TAG_TRIANGLE; s.payload.tri.base = i % 20 + 1; s.payload.tri.height = i % 40 + 1; break;
        case 3: s.tag = TAG_POINT;    break;
        case 4: s.tag = TAG_LINE;     s.payload.length = i % 100 + 1; break;
        default: s.tag = TAG_RECT;    s.payload.rect.w = i % 25 + 1; s.payload.rect.h = i % 35 + 1; break;
    }
    return s;
}

static double tv_to_sec(const struct timeval *tv) {
    return (double)tv->tv_sec + (double)tv->tv_usec / 1e6;
}

static double ts_to_sec(const struct timespec *ts) {
    return (double)ts->tv_sec + (double)ts->tv_nsec / 1e9;
}

int main(void) {
    struct timespec wall0, wall1;
    struct rusage ru0, ru1;

    getrusage(RUSAGE_SELF, &ru0);
    clock_gettime(CLOCK_MONOTONIC, &wall0);

    int64_t total = 0;
    for (int64_t i = 0; i < 100000; i++) {
        total += area(make_shape(i));
    }

    clock_gettime(CLOCK_MONOTONIC, &wall1);
    getrusage(RUSAGE_SELF, &ru1);

    double wall = ts_to_sec(&wall1) - ts_to_sec(&wall0);
    double cpu = (tv_to_sec(&ru1.ru_utime) - tv_to_sec(&ru0.ru_utime))
               + (tv_to_sec(&ru1.ru_stime) - tv_to_sec(&ru0.ru_stime));
    double peak_kb = (double)ru1.ru_maxrss;

    printf("checksum = %lld\n", (long long)total);
    printf("__BENCH_METRICS__\n");
    printf("wall_time_s=%.6f\n", wall);
    printf("cpu_time_s=%.6f\n", cpu);
    printf("peak_memory_kb=%.1f\n", peak_kb);
    return 0;
}

File: benchmarks/cross_language/go/enum_match.go
// Benchmark: Enum dispatch -- Go equivalent.
// Tagged union via discriminant int + two payload fields.
// Mirrors the Mapanare 6-variant Shape (Circle, Square, Triangle, Point, Line, Rect).
// Checksum must equal 52818168 across all languages.
package main

import (
	"fmt"
	"runtime"
	"syscall"
	"time"
)

type Shape struct {
	Tag int64 // 0=Circle, 1=Square, 2=Triangle, 3=Point, 4=Line, 5=Rect
	A   int64
	B   int64
}

func area(s Shape) int64 {
	switch s.Tag {
	case 0:
		return s.A * s.A * 3
	case 1:
		return s.A * s.A
	case 2:
		return s.A * s.B / 2
	case 3:
		return 0
	case 4:
		return s.A
	default: // 5 = Rect
		return s.A * s.B
	}
}

func makeShape(i int64) Shape {
	tag := i % 6
	switch tag {
	case 0:
		return Shape{Tag: 0, A: i%50 + 1}
	case 1:
		return Shape{Tag: 1, A: i%30 + 1}
	case 2:
		return Shape{Tag: 2, A: i%20 + 1, B: i%40 + 1}
	case 3:
		return Shape{Tag: 3}
	case 4:
		return Shape{Tag: 4, A: i%100 + 1}
	default:
		return Shape{Tag: 5, A: i%25 + 1, B: i%35 + 1}
	}
}

func main() {
	runtime.GC()
	var ruStart, ruEnd syscall.Rusage
	syscall.Getrusage(syscall.RUSAGE_SELF, &ruStart)
	wall0 := time.Now()

	var total int64
	for i := int64(0); i < 100000; i++ {
		total += area(makeShape(i))
	}

	wall := time.Since(wall0).Seconds()
	syscall.Getrusage(syscall.RUSAGE_SELF, &ruEnd)
	cpu := timevalSub(ruEnd.Utime, ruStart.Utime) + timevalSub(ruEnd.Stime, ruStart.Stime)
	peakKB := float64(ruEnd.Maxrss)

	fmt.Printf("checksum = %d\n", total)
	fmt.Println("__BENCH_METRICS__")
	fmt.Printf("wall_time_s=%.6f\n", wall)
	fmt.Printf("cpu_time_s=%.6f\n", cpu)
	fmt.Printf("peak_memory_kb=%.1f\n", peakKB)
}

func timevalSub(end, start syscall.Timeval) float64 {
	return float64(end.Sec-start.Sec) + float64(end.Usec-start.Usec)/1e6
}
